File: go-server/browser_registry.go
package main

import (
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"sync"
)

var ErrBrowserRegistryConflict = errors.New("browser registration conflict")

type BrowserRegistration struct {
	ConnectID string `json:"connect_id"`
	Name      string `json:"name"`
}

type browserRegistryFile struct {
	Browsers []BrowserRegistration `json:"browsers"`
}

type BrowserRegistry struct {
	mu       sync.RWMutex
	path     string
	browsers map[string]string
}

func NewBrowserRegistry(path string) (*BrowserRegistry, error) {
	if path == "" {
		path = filepath.Join(".", "browser_registry.json")
	}

	registry := &BrowserRegistry{
		path:     path,
		browsers: make(map[string]string),
	}
	if err := registry.load(); err != nil {
		return nil, err
	}
	return registry, nil
}

func (br *BrowserRegistry) load() error {
	data, err := os.ReadFile(br.path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil
		}
		return err
	}

	var file browserRegistryFile
	if err := json.Unmarshal(data, &file); err != nil {
		return err
	}

	for _, browser := range file.Browsers {
		connectID := strings.TrimSpace(browser.ConnectID)
		name := strings.TrimSpace(browser.Name)
		if connectID != "" && name != "" {
			br.browsers[connectID] = name
		}
	}
	return nil
}

func (br *BrowserRegistry) save() error {
	if err := os.MkdirAll(filepath.Dir(br.path), 0o755); err != nil {
		return err
	}

	ids := make([]string, 0, len(br.browsers))
	for connectID := range br.browsers {
		ids = append(ids, connectID)
	}
	sort.Strings(ids)

	file := browserRegistryFile{Browsers: make([]BrowserRegistration, 0, len(ids))}
	for _, connectID := range ids {
		file.Browsers = append(file.Browsers, BrowserRegistration{
			ConnectID: connectID,
			Name:      br.browsers[connectID],
		})
	}

	data, err := json.MarshalIndent(file, "", "  ")
	if err != nil {
		return err
	}
	data = append(data, '\n')

	tmpPath := br.path + ".tmp"
	if err := os.WriteFile(tmpPath, data, 0o644); err != nil {
		return err
	}
	return os.Rename(tmpPath, br.path)
}

func (br *BrowserRegistry) Register(connectID, name string) (BrowserRegistration, error) {
	connectID = strings.TrimSpace(connectID)
	name = strings.TrimSpace(name)
	if connectID == "" {
		return BrowserRegistration{}, errors.New("connect_id is required")
	}
	if name == "" {
		return BrowserRegistration{}, errors.New("name is required")
	}

	br.mu.Lock()
	defer br.mu.Unlock()

	if existingName, ok := br.browsers[connectID]; ok {
		if existingName == name {
			return BrowserRegistration{ConnectID: connectID, Name: name}, nil
		}
		return BrowserRegistration{}, ErrBrowserRegistryConflict
	}

	for existingID, existingName := range br.browsers {
		if existingID != connectID && existingName == name {
			return BrowserRegistration{}, ErrBrowserRegistryConflict
		}
	}

	br.browsers[connectID] = name
	if err := br.save(); err != nil {
		delete(br.browsers, connectID)
		return BrowserRegistration{}, err
	}

	return BrowserRegistration{ConnectID: connectID, Name: name}, nil
}

func (br *BrowserRegistry) Validate(connectID, name string) bool {
	br.mu.RLock()
	defer br.mu.RUnlock()
	return br.browsers[connectID] == name
}

func (br *BrowserRegistry) GetName(connectID string) (string, bool) {
	br.mu.RLock()
	defer br.mu.RUnlock()
	name, ok := br.browsers[connectID]
	return name, ok
}

File: python-server/browser_registry.py
"""Persistent browser registration storage."""

from __future__ import annotations

import json
from pathlib import Path
from threading import RLock
from typing import Dict, List, Optional


class BrowserRegistryError(ValueError):
    """Raised when a browser registration request is invalid."""


class BrowserRegistry:
    def __init__(self, path: Optional[Path] = None):
        self.path = path or Path(__file__).with_name("browser_registry.json")
        self._lock = RLock()
        self._browsers: Dict[str, str] = {}
        self._load()

    def _load(self) -> None:
        if not self.path.exists():
            return

        with self.path.open("r", encoding="utf-8") as file:
            data = json.load(file)

        browsers = data.get("browsers", []) if isinstance(data, dict) else []
        self._browsers = {
            str(item["connect_id"]): str(item["name"])
            for item in browsers
            if item.get("connect_id") and item.get("name")
        }

    def _save(self) -> None:
        self.path.parent.mkdir(parents=True, exist_ok=True)
        data = {
            "browsers": [
                {"connect_id": connect_id, "name": name}
                for connect_id, name in sorted(self._browsers.items())
            ]
        }
        tmp_path = self.path.with_suffix(self.path.suffix + ".tmp")
        with tmp_path.open("w", encoding="utf-8") as file:
            json.dump(data, file, ensure_ascii=False, indent=2)
            file.write("\n")
        tmp_path.replace(self.path)

    def register(self, connect_id: str, name: str) -> Dict[str, str]:
        connect_id = connect_id.strip()
        name = name.strip()
        if not connect_id:
            raise BrowserRegistryError("connect_id is required")
        if not name:
            raise BrowserRegistryError("name is required")

        with self._lock:
            existing_name = self._browsers.get(connect_id)
            if existing_name is not None:
                if existing_name == name:
                    return {"connect_id": connect_id, "name": name}
                raise BrowserRegistryError("connect_id is already registered with a different name")

            for existing_id, existing in self._browsers.items():
                if existing == name and existing_id != connect_id:
                    raise BrowserRegistryError("name is already registered")

            self._browsers[connect_id] = name
            self._save()
            return {"connect_id": connect_id, "name": name}

    def validate(self, connect_id: str, name: str) -> bool:
        with self._lock:
            return self._browsers.get(connect_id) == name

    def get_name(self, connect_id: str) -> Optional[str]:
        with self._lock:
            return self._browsers.get(connect_id)

    def list_registered(self) -> List[Dict[str, str]]:
        with self._lock:
            return [
                {"connect_id": connect_id, "name": name}
                for connect_id, name in sorted(self._browsers.items())
            ]
